hw/ip/aes/dv/aes_model_dpi/aes_model_dpi.c

```diff
@@ -106,6 +107,98 @@ void c_dpi_aes_crypt(const unsigned char impl_i, const unsigned char op_i,
   return;
 }
 
+void c_dpi_aes_crypt_message(unsigned char impl_i, unsigned char op_i,
+                             const svBitVecVal *mode_i, const svBitVecVal *iv_i,
+                             const svBitVecVal *key_len_i,
+                             const svBitVecVal *key_i,
+                             const svOpenArrayHandle data_i,
+                             svOpenArrayHandle data_o) {
+  if (impl_i == 0) {
+    // The C model is currently not supported.
+    printf(
+        "ERROR: c_dpi_aes_crypt_message() currently supports OpenSSL/BoringSSL "
+        "only\n");
+    return;
+  }
+
+  // Get key from simulator.
+  unsigned char *key = aes_key_get(key_i);
+
+  // key_len_i is one-hot encoded.
+  int key_len;
+  if (*key_len_i == 0x1) {
+    key_len = 16;
+  } else if (*key_len_i == 0x2) {
+    key_len = 24;
+  } else {  // 0x4
+    key_len = 32;
+  }
+
+  crypto_mode_t mode = (crypto_mode_t)*mode_i;
+
+  // Modes other than ECB require an IV from the simulator.
+  unsigned char *iv = (unsigned char *)malloc(16 * sizeof(unsigned char));
+  if (!iv) {
+    printf("ERROR: malloc() for c_dpi_aes_crypt_message failed");
+    return;
```

Comment:
General note: if you're going for error handling for allocation failures, you should also ensure to call the free() functions. Typically that's done with a `goto` jumping to the end. 

Alternatively (and that's what I would do), avoid handling assertion errors and simply assert if it fails. In most cases, there's nothing you can do anyways. (That's obviously not true for huge allocations, but that's not applicable here.)

Reply:
That is a good point. I switched to `assert()`.